File: info_interattiva.rb
#!/usr/bin/env ruby

require 'net/http'
require 'json'
require 'uri'

class GitHubInfo
  def initialize(username = 'mario Rossi')
    @username = username
    @url = "https://api.github.com/users/#{username}"
  end

  def get_info
    begin
      uri = URI(@url)
      response = Net::HTTP.get_response(uri)
      if response.code == '200'
        data = JSON.parse(response.body)
        {
          name: data['name'] || 'N/A',
          bio: data['bio'] || 'N/A',
          public_repos: data['public_repos'],
          followers: data['followers'],
          following: data['following']
        }
      else
        { error: "Utente non trovato o errore nella richiesta API" }
      end
    rescue => e
      { error: "Errore durante il recupero delle informazioni: #{e.message}" }
    end
  end
end

def mostra_menu
  puts "\n" + "="*50
  puts "Seleziona un'opzione:"
  puts "1. Nome"
  puts "2. Bio"
  puts "3. Numero di repository pubblici"
  puts "4. Numero di follower"
  puts "5. Numero di following"
  puts "6. Tutte le informazioni"
  puts "7. Esci"
  print "Inserisci il numero dell'opzione: "
end

def main
  puts "Benvenuto nell'applicazione informativa GitHub!"
  print "Inserisci il nome utente GitHub da cercare (default: mario Rossi): "
  input_username = gets.chomp.strip
  username = input_username.empty? ? 'mario Rossi' : input_username
  
  github_info = GitHubInfo.new(username)
  info = github_info.get_info
  
  if info[:error]
    puts info[:error]
    return
  end

  loop do
    mostra_menu
    scelta = gets.chomp.to_i

    case scelta
    when 1
      puts "\nNome: #{info[:name]}"
    when 2
      puts "\nBio: #{info[:bio]}"
    when 3
      puts "\nNumero di repository pubblici: #{info[:public_repos]}"
    when 4
      puts "\nNumero di follower: #{info[:followers]}"
    when 5
      puts "\nNumero di following: #{info[:following]}"
    when 6
      puts "\nTutte le informazioni:"
      puts "Nome: #{info[:name]}"
      puts "Bio: #{info[:bio]}"
      puts "Repository pubblici: #{info[:public_repos]}"
      puts "Follower: #{info[:followers]}"
      puts "Following: #{info[:following]}"
    when 7
      puts "\nGrazie per aver usato l'applicazione. Arrivederci!"
      break
    else
      puts "\nOpzione non valida. Riprova."
    end
    
    puts "\nPremi Invio per continuare..."
    gets
  end
end

main

File: info_interattiva.pl
#!/usr/bin/swipl

% Questo file Prolog implementa un'interfaccia interattiva per visualizzare informazioni da GitHub
% tramite una simulazione del comportamento richiesto, poiché Prolog non gestisce direttamente
% le richieste HTTP senza estensioni specifiche. Le informazioni sono simulate per dimostrare
% l'interfaccia interattiva richiesta.

:- initialization(main).

% Predicati per le informazioni simulate (in una vera implementazione questi verrebbero da una chiamata API)
nome('Mario Rossi').
bio('Sono uno sviluppatore appassionato di logica e programmazione funzionale.').
repo_pubblici(15).
follower(128).
following(50).

% Mostra il menu principale
mostra_menu :-
    write('=================================================='), nl,
    write('Seleziona un\'opzione:'), nl,
    write('1. Nome'), nl,
    write('2. Bio'), nl,
    write('3. Numero di repository pubblici'), nl,
    write('4. Numero di follower'), nl,
    write('5. Numero di following'), nl,
    write('6. Tutte le informazioni'), nl,
    write('7. Esci'), nl,
    write('Inserisci il numero dell\'opzione: ').

% Gestione delle scelte
esegui_scelta(1) :-
    nome(Nome),
    format('~nNome: ~w~n', [Nome]).
esegui_scelta(2) :-
    bio(Bio),
    format('~nBio: ~w~n', [Bio]).
esegui_scelta(3) :-
    repo_pubblici(Repos),
    format('~nNumero di repository pubblici: ~w~n', [Repos]).
esegui_scelta(4) :-
    follower(Followers),
    format('~nNumero di follower: ~w~n', [Followers]).
esegui_scelta(5) :-
    following(Following),
    format('~nNumero di following: ~w~n', [Following]).
esegui_scelta(6) :-
    nome(Nome),
    bio(Bio),
    repo_pubblici(Repos),
    follower(Followers),
    following(Following),
    write('Tutte le informazioni:'), nl,
    format('Nome: ~w~n', [Nome]),
    format('Bio: ~w~n', [Bio]),
    format('Repository pubblici: ~w~n', [Repos]),
    format('Follower: ~w~n', [Followers]),
    format('Following: ~w~n', [Following]).
esegui_scelta(7) :-
    write('Grazie per aver usato l\'applicazione. Arrivederci!'), nl,
    halt.
esegui_scelta(_) :-
    write('Opzione non valida. Riprova.'), nl.

% Legge l'input dell'utente
leggi_input(Scelta) :-
    read(Scelta).

% Loop principale
loop :-
    mostra_menu,
    leggi_input(Scelta),
    esegui_scelta(Scelta),
    write('Premi Invio per continuare...'), nl,
    get_char(_),
    loop.

% Punto di ingresso principale
main :-
    write('Benvenuto nell\'applicazione informativa GitHub!'),
    nl,
    write('ATTENZIONE: Questo programma simula le informazioni di GitHub.'),
    nl,
    write('In una vera implementazione, questo codice farebbe una chiamata API.'),
    nl,
    loop.